partitioner/src/main/java/io/confluent/connect/storage/partitioner/TimeBasedPartitioner.java

```diff
@@ -17,31 +17,47 @@
 package io.confluent.connect.storage.partitioner;
 
 import org.apache.kafka.common.config.ConfigException;
+import org.apache.kafka.connect.connector.ConnectRecord;
+import org.apache.kafka.connect.data.Schema;
+import org.apache.kafka.connect.data.Schema.Type;
+import org.apache.kafka.connect.data.Struct;
+import org.apache.kafka.common.utils.Time;
+import org.apache.kafka.connect.errors.ConnectException;
 import org.apache.kafka.connect.sink.SinkRecord;
 import org.joda.time.DateTime;
 import org.joda.time.DateTimeZone;
 import org.joda.time.format.DateTimeFormat;
 import org.joda.time.format.DateTimeFormatter;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 import java.lang.reflect.InvocationTargetException;
 import java.util.Locale;
 import java.util.Map;
 
 import io.confluent.connect.storage.common.SchemaGenerator;
 import io.confluent.connect.storage.common.StorageCommonConfig;
+import io.confluent.connect.storage.errors.PartitionException;
 
 public class TimeBasedPartitioner<T> extends DefaultPartitioner<T> {
   // Duration of a partition in milliseconds.
+  private static final Logger log = LoggerFactory.getLogger(TimeBasedPartitioner.class);
   private long partitionDurationMs;
+  private String pathFormat;
   private DateTimeFormatter formatter;
+  protected TimestampExtractor timestampExtractor;
 
   protected void init(long partitionDurationMs, String pathFormat, Locale locale, DateTimeZone timeZone,
                       Map<String, Object> config) {
     delim = (String) config.get(StorageCommonConfig.DIRECTORY_DELIM_CONFIG);
     this.partitionDurationMs = partitionDurationMs;
+    this.pathFormat = pathFormat;
     this.formatter = getDateTimeFormatter(pathFormat, timeZone).withLocale(locale);
     try {
       partitionFields = newSchemaGenerator(config).newPartitionFields(pathFormat);
+      timestampExtractor = newTimestampExtractor(
+          (String) config.get(PartitionerConfig.TIMESTAMP_EXTRACTOR_CLASS_CONFIG));
+      timestampExtractor.configure(config);
```

Comment:
Coming back to the "it's not a class" comment above, if we have the same structure, is there any reason *not* to just make it a pluggable class?

The reason I ask is that even fields may not be ideal, e.g. if your timestamp is embedded deeper in the message, e.g. 
```
{
  "header": {
    "timestamp": 100,
    "dc": "dc1",
  },
  "payload": { ... }
}
```
 Of course with an SMT you can always pull that up to a top-level field and building too much flexibility into extracting the timestamp might be undesirable, but then you could also just use  an SMT to set the record's timestamp (with more flexibility, e.g. using dotted notation) and then only offer wallclock or record?